// react-dnd-preview/dist/esm/usePreview.js
import { useRef } from 'react';
import { useDragLayer } from 'react-dnd';
import { calculatePointerPosition } from './offsets';

var getStyle = function getStyle(currentOffset) {
  var transform = "translate(".concat(currentOffset.x, "px, ").concat(currentOffset.y, "px)");
  return {
    pointerEvents: 'none',
    position: 'fixed',
    top: 0,
    left: 0,
    transform: transform,
    WebkitTransform: transform
  };
};

export var usePreview = function usePreview() {
  var child = useRef(null);
  var collectedProps = useDragLayer(function (monitor) {
    return {
      currentOffset: calculatePointerPosition(monitor, child),
      isDragging: monitor.isDragging(),
      itemType: monitor.getItemType(),
      item: monitor.getItem(),
      monitor: monitor
    };
  });

  if (!collectedProps.isDragging || collectedProps.currentOffset === null) {
    return {
      display: false
    };
  }

  return {
    display: true,
    itemType: collectedProps.itemType,
    item: collectedProps.item,
    style: getStyle(collectedProps.currentOffset),
    monitor: collectedProps.monitor,
    ref: child
  };
};
