// dnd-multi-backend/dist/cjs/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "MultiBackend", {
  enumerable: true,
  get: function get() {
    return _MultiFactory.MultiFactory;
  }
});
Object.defineProperty(exports, "createTransition", {
  enumerable: true,
  get: function get() {
    return _createTransition.createTransition;
  }
});
Object.defineProperty(exports, "HTML5DragTransition", {
  enumerable: true,
  get: function get() {
    return _transitions.HTML5DragTransition;
  }
});
Object.defineProperty(exports, "TouchTransition", {
  enumerable: true,
  get: function get() {
    return _transitions.TouchTransition;
  }
});
Object.defineProperty(exports, "MouseTransition", {
  enumerable: true,
  get: function get() {
    return _transitions.MouseTransition;
  }
});
Object.defineProperty(exports, "PointerTransition", {
  enumerable: true,
  get: function get() {
    return _transitions.PointerTransition;
  }
});

var _MultiFactory = require("./MultiFactory");

var _createTransition = require("./createTransition");

var _transitions = require("./transitions");
