// react-dnd/lib/hooks/useDrag/DragSourceImpl.js
import { invariant } from '@react-dnd/invariant';
export class DragSourceImpl {
    constructor(spec, monitor, connector) {
        this.spec = spec;
        this.monitor = monitor;
        this.connector = connector;
    }
    beginDrag() {
        const spec = this.spec;
        const monitor = this.monitor;
        const { begin, item } = spec;
        if (begin) {
            const beginResult = begin(monitor);
            invariant(beginResult == null || typeof beginResult === 'object', 'dragSpec.begin() must either return an object, undefined, or null');
            return beginResult ?? item ?? {};
        }
        return item ?? {};
    }
    canDrag() {
        const spec = this.spec;
        const monitor = this.monitor;
        if (typeof spec.canDrag === 'boolean') {
            return spec.canDrag;
        }
        else if (typeof spec.canDrag === 'function') {
            return spec.canDrag(monitor);
        }
        else {
            return true;
        }
    }
    isDragging(globalMonitor, target) {
        const spec = this.spec;
        const monitor = this.monitor;
        const { isDragging } = spec;
        return isDragging
            ? isDragging(monitor)
            : target === globalMonitor.getSourceId();
    }
    endDrag() {
        const spec = this.spec;
        const monitor = this.monitor;
        const connector = this.connector;
        const { end } = spec;
        if (end) {
            end(monitor.getItem(), monitor);
        }
        connector.reconnect();
    }
}
