// react-dnd-preview/dist/cjs/usePreview.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.usePreview = void 0;

var _react = require("react");

var _reactDnd = require("react-dnd");

var _offsets = require("./offsets");

var getStyle = function getStyle(currentOffset) {
  var transform = "translate(".concat(currentOffset.x, "px, ").concat(currentOffset.y, "px)");
  return {
    pointerEvents: 'none',
    position: 'fixed',
    top: 0,
    left: 0,
    transform: transform,
    WebkitTransform: transform
  };
};

var usePreview = function usePreview() {
  var child = (0, _react.useRef)(null);
  var collectedProps = (0, _reactDnd.useDragLayer)(function (monitor) {
    return {
      currentOffset: (0, _offsets.calculatePointerPosition)(monitor, child),
      isDragging: monitor.isDragging(),
      itemType: monitor.getItemType(),
      item: monitor.getItem(),
      monitor: monitor
    };
  });

  if (!collectedProps.isDragging || collectedProps.currentOffset === null) {
    return {
      display: false
    };
  }

  return {
    display: true,
    itemType: collectedProps.itemType,
    item: collectedProps.item,
    style: getStyle(collectedProps.currentOffset),
    monitor: collectedProps.monitor,
    ref: child
  };
};

exports.usePreview = usePreview;
